core/src/sumeragi/main_loop.rs

```diff
@@ -0,0 +1,1340 @@
+    pub network: Addr<IrohaNetwork>,
+    /// [`PhantomData`] used to generify over [`FaultInjection`] implementations
+    pub fault_injection: PhantomData<F>, // TODO: remove
+    /// The size of batch that is being gossiped. Smaller size leads
+    /// to longer time to synchronise, useful if you have high packet loss.
+    pub gossip_batch_size: u32,
+    /// The time between gossiping. More frequent gossiping shortens
+    /// the time to sync, but can overload the network.
+    pub gossip_period: Duration,
+    /// [`PeerId`]s of the peers that are currently online.
+    pub current_online_peers: Mutex<Vec<PeerId>>,
+    /// Hash of the latest block
+    pub latest_block_hash: Mutex<HashOf<VersionedCommittedBlock>>,
+    /// Incoming?? sender channel
+    pub incoming_message_sender: Mutex<mpsc::SyncSender<MessagePacket>>,
+    /// Incoming message receiver channel.
+    pub incoming_message_receiver: Mutex<mpsc::Receiver<MessagePacket>>,
+}
+
+impl<F: FaultInjection> Debug for SumeragiWithFault<F> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Sumeragi")
+            .field("public_key", &self.key_pair.public_key())
+            .field("peer_id", &self.peer_id)
+            .finish()
+    }
+}
+
+// TODO: In general naming things after the programming patterns is
+// considered a bad practice. We need a better name.
+/// Internal structure that retains the state.
+pub struct State {
+    /// The [`GenesisNetwork`] that was used to initialise the state machine.
+    pub genesis_network: Option<GenesisNetwork>,
+    /// The hash of the latest [`VersionedCommittedBlock`]
+    pub latest_block_hash: HashOf<VersionedCommittedBlock>,
+    /// Current block height
+    pub latest_block_height: u64,
+    /// The current network topology.
+    pub current_topology: Topology,
+    /// The sumeragi internal `WorldStateView`. This will probably
+    /// morph into a wsv + various patches as we attempt to
+    /// multithread isi execution. In the future we might also once
+    /// again merge the internal wsv with the public facing one. But
+    /// as of now we keep them seperate for greater flexibility when
+    /// optimizing.
+    pub wsv: WorldStateView,
+    /// In order to *be fast*, we must minimize communication with
+    /// other subsystems where we can. This way the performance of
+    /// sumeragi is more dependent on the code that is internal to the
+    /// subsystem.
+    ///
+    /// This transaction cache was therefore introduced so that only
+    /// interact with the transaction queue as necessary. If this were
+    /// written in C it would have just been an array of
+    /// `VersionedAcceptedTransaction`, but because of Rust's RAII
+    /// enforcement it has to be an array of options.  Otherwise we
+    /// would need to use unsafe to implement the fast pruning on the
+    /// array.
+    pub transaction_cache: Vec<Option<VersionedAcceptedTransaction>>,
```

Comment:
Especially then AFAIK another good approaches without `Option` and `unsafe` were disscussed

Reply:
This comment actually has a point. In C, this approach would have largely worked, except in some cases it would have led to an occasional segfault. More importantly, this doesn't actually take up more space, (remember the niche optimisation).